onnxruntime/contrib_ops/cpu/nchwc_ops.cc

```diff
@@ -0,0 +1,222 @@
+    NchwcAveragePool);
+
+ONNX_CPU_OPERATOR_TYPED_NCHWC_KERNEL(
+    GlobalAveragePool,
+    1,
+    float,
+    KernelDefBuilder()
+        .TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
+    NchwcAveragePool);
+
+template <typename T>
+Status ReorderInput<T>::Compute(OpKernelContext* context) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  const TensorShape& X_shape = X->Shape();
+  ORT_ENFORCE(X_shape.NumDimensions() == 4);
+  ORT_ENFORCE((X_shape[1] % MlasNchwcGetBlockSize()) == 0);
+  Tensor* Y = context->Output(0, X_shape);
+  MlasReorderInput(X_shape.GetDims().data(), X->template Data<T>(), Y->template MutableData<T>());
+  return Status::OK();
+}
+
+template <typename T>
+Status ReorderOutput<T>::Compute(OpKernelContext* context) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  const TensorShape& X_shape = X->Shape();
+  ORT_ENFORCE(X_shape.NumDimensions() == 4);
+  std::vector<int64_t> Y_shape(X_shape.GetDims());
+  ORT_ENFORCE(channels_ <= Y_shape[1]);
+  Y_shape[1] = channels_;
+  Tensor* Y = context->Output(0, Y_shape);
+  MlasReorderOutput(Y_shape.data(), X->template Data<T>(), Y->template MutableData<T>());
+  return Status::OK();
+}
+
+template <typename T>
+Status NchwcConv<T>::Compute(OpKernelContext* context) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  const Tensor* W = context->Input<Tensor>(1);
+  const Tensor* B = context->Input<Tensor>(2);
+  const Tensor* Sum = context->Input<Tensor>(3);
+
+  ORT_RETURN_IF_ERROR(ConvBase::ValidateInputShape(X, W));
+
+  const TensorShape& X_shape = X->Shape();
+  const TensorShape& W_shape = W->Shape();
+  ORT_ENFORCE(X_shape.NumDimensions() == 4);
+
+  const size_t nchwc_block_size = MlasNchwcGetBlockSize();
+  ORT_ENFORCE((static_cast<size_t>(X_shape[1]) < nchwc_block_size) || ((X_shape[1] % nchwc_block_size) == 0));
+
+  std::vector<int64_t> kernel_shape;
+  ORT_RETURN_IF_ERROR(ConvBase::ComputeKernelShape(W_shape, kernel_shape));
+  if (kernel_shape.size() != 2) {
+    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT, "Unsupported convolution size.");
+  }
+
+  std::vector<int64_t> pads(ConvBase::pads_);
+  if (pads.empty()) {
+    pads.resize(kernel_shape.size() * 2, 0);
+  }
+  std::vector<int64_t> dilations(ConvBase::dilations_);
+  if (dilations.empty()) {
+    dilations.resize(kernel_shape.size(), 1);
+  }
+  std::vector<int64_t> strides(ConvBase::strides_);
+  if (strides.empty()) {
+    strides.resize(kernel_shape.size(), 1);
+  }
+
+  std::vector<int64_t> Y_dims;
+  Y_dims.insert(Y_dims.begin(), {X_shape[0], W_shape[0]});
+  TensorShape input_shape = X->Shape().Slice(2);
+  ORT_RETURN_IF_ERROR(ConvBase::InferOutputShape(input_shape, kernel_shape, strides, dilations, &pads, &Y_dims));
+  Tensor* Y = context->Output(0, Y_dims);
+  T* y_data = Y->template MutableData<T>();
+
+  // Check for the optional Conv/Sum fusion.
+  if (Sum != nullptr) {
+    const auto& sum_shape = Sum->Shape();
```

Comment:
How about this: Adding a new function: CopyTo() to the Tensor class, move the code here to that function?

Reply:
I'd rather not add something to Tensor for this PR.